README.md
===========================================
pyprimes — generate and test prime numbers
===========================================


Introduction
------------

Prime numbers are those positive integers which are not divisible exactly
by any number other than itself or one. Generating primes and testing for
primality has been a favourite mathematical pastime for centuries, as well
as of great practical importance for encrypting data. The ``pyprimes``
package provides functions for generating prime numbers on the fly and
testing whether or not a given number is prime.

Features include:

- Produce prime numbers lazily, on demand.
- Effective, fast algorithms including Sieve of Eratosthenes,
  Croft Spiral, and Wheel Factorisation.
- Test whether numbers are prime efficiently.
- Deterministic and probabilistic primality tests.
- Examples of what *not* to do provided, including naive trial
  division, Turner's algorithm (Euler's sieve), and primality
  testing using a regular expression.
- Factorise small numbers into the product of prime factors.
- Suitable for Python 2.4 through 3.x from one code base.


Installation
------------

To install, extract the tarball into the current directory, then cd into the
expanded directory. Run:

    python setup.py install

from your system shell (not the python interpreter) to install.


Licence
-------

pyprimes is licenced under the MIT Licence. See the ``LICENCE.txt`` file and the
header of ``pyprimes/__init__.py``.


Test suite
----------

pyprimes comes with an extensive test suite containing unit tests, doc tests
and regression tests. To run the test suite, run the ``pyprimes/tests.py``
module.

For Python version 2.5 on up, the most convenient way to do this after
installation is from your system shell:

    python -m pyprimes.test

To get even more verbose output, pass the -v switch:

    python -m pyprimes.test -v

The -m switch is not supported by Python 2.4, to run the test suite under
that version you will have to give the path to the test module and run it
manually. E.g. this may work on some systems:

    cd /usr/lib/python2.4
    python2.4 pyprimes/test.py


Known Issues and Bugs
---------------------

1) The API of this package is not yet stable. That means that the names of
   functions, the arguments that they take, values returned, etc. may
   change without warning in the future.

2) With older versions of Python (2.4 - 2.6), doctests are not run when
   running the test suite. You can run them manually for each individual
   module, e.g.:

    python2.6 -m doctest pyprimes/factors.py -v

3) In general, prime-related functions expect their numeric arguments to
   be integers, but don't perform any type-checking. If you pass a float,
   Decimal, Fraction or other non-int (or long) value, behaviour is
   undefined.

4) The package is currently rather inconsistent in whether it uses
   floating point sqrt or integer-valued sqrt. This should be considered
   a known bug.



CHANGES.txt
=========================================================================
                           C H A N G E   L O G
=========================================================================

Version 0.2.2a 2015-01-19
    * Improved handling of expensive tests (thanks to Ben Finney)

Version 0.2.1a 2015-01-08
    * Version bump to satisfy PyPI packaging difficulties.

Version 0.2a 2015-01-07
    * Python 3.x support.
    * pyprimes now also works with Python 2.4 through 2.7.
    * Refactored to a package layout instead of a single file.
    * Tests and compatibility layer are documented as private
      implementation details.
    * Test suite has significantly more coverage.
    * primes() accepts optional start and end values.
    * primes_above() and primes_below() are redundant and removed in
      favour of primes(start, end).
    * Support for generating and testing primes with alternate
      implementations or algorithms (pyprimes.strategic module).
    * Much improved documentation for probabilistic algorithms.
    * Primality testing supports probabilistic as well as deterministic
      tests, that is, "not prime, certainly prime, probably prime".
    * Added prev_prime() function.
    * Swapped names for factors() and factorise().
    * Fast, efficient deterministic primality tests up to 2**64.

Version 0.1.3a 2014-09-11 (not publicly released)
    * Add ability to generate primes between start and end limits.
    * Add a quite naive next_prime function.
    * Add test suite.

Version 0.1.2a 2012-08-25
    * Move "awful" and naive algorithms into their own namespace.
    * Improve documentation for such awful algorithms.
    * "Probably prime" warnings now default to off.
    * Use an explicit function argument instead of a global variable
      for "probably prime" warnings.
    * Allow probabilistic primality tests to set the number of trials.

Version 0.1.1a 2012-02-22
    * Work around mbcs bug in distutils: see issue 10945 on the Python
      bug tracker http://bugs.python.org/issue10945
    * Fixed other minor issues with the setup script and metadata.
    * Neglected to update this changelog, sigh.

Version 0.1.0 2012-02-22
    Initial release.



src/pyprimes/__init__.py
# -*- coding: utf-8 -*-

##  Package pyprimes.py
##
##  Copyright © 2015 Steven D'Aprano.
##
##  Permission is hereby granted, free of charge, to any person obtaining
##  a copy of this software and associated documentation files (the
##  "Software"), to deal in the Software without restriction, including
##  without limitation the rights to use, copy, modify, merge, publish,
##  distribute, sublicense, and/or sell copies of the Software, and to
##  permit persons to whom the Software is furnished to do so, subject to
##  the following conditions:
##
##  The above copyright notice and this permission notice shall be
##  included in all copies or substantial portions of the Software.
##
##  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
##  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
##  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
##  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
##  CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
##  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
##  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

"""\
================
Package pyprimes
================

This package includes functions for generating prime numbers, primality
testing, and factorising numbers into prime factors.


Definitions
-----------

"Prime numbers" are positive integers with no factors other than themselves
and 1. The first few primes are 2, 3, 5, 7, 11, ... . There is only one
even prime number, namely 2, and an infinite number of odd prime numbers.

"Composite numbers" are positive integers which have factors other than
themselves and 1. Composite numbers can be uniquely factorised into the
product of two or more (possibly repeated) primes, e.g. 18 = 2*3*3.

Both 0 and 1 are considered by mathematicians to be neither prime nor
composite.


Generating prime numbers
------------------------

To generate an unending stream of prime numbers, use the ``primes``
function with no arguments:

    >>> p = primes()
    >>> [next(p) for _ in range(10)]
    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

Or give start and end arguments for just a subset of the primes:

    >>> list(primes(180, 225))
    [181, 191, 193, 197, 199, 211, 223]


The ``next_prime`` and ``prev_prime`` functions return the next, or
previous, prime from the given argument:

    >>> next_prime(20)
    23
    >>> prev_prime(20)
    19

    NOTE:: For large prime numbers p, the *average* distance between p and
           the next (or previous) prime is proportional to log p. If p
           is large enough, it may take some considerable time to calculate
           the next or previous prime from p.


Primality testing
-----------------

To test whether an integer is prime or not, use ``is_prime``:

    >>> is_prime(10)
    False
    >>> is_prime(11)
    True

For extremely large values ``is_prime`` may be probabilistic. That is,
if it reports a number is prime, it may be only "almost certainly prime",
with a very small chance that the number is actually composite. If it
returns False, the number is certainly composite. For more details on the
probabilistic nature of primality testing, see the ``probabilistic`` module.

The ``trial_division`` function also returns True for primes and False for
non-primes, unlike ``is_prime`` it is always an exact test and never
deterministic. However, it may be very slow and require large amounts of
memory for very large values.

    >>> trial_division(15)
    False
    >>> trial_division(17)
    True


Number theory convenience functions
-----------------------------------

``pyprimes`` offers a few convenience functions from the Number Theory
branch of mathematics.

    nprimes:
        Return the first n primes.

    nth_prime:
        Return the nth prime.

    prime_count:
        Return the number of primes less than n.

    prime_sum:
        Return the sum of primes less than n.

    prime_partial_sums:
        Yield the running sums of primes less than n.

See the individual functions for further details.


Sub-modules
-----------

The ``pyprimes`` package also includes the following public sub-modules:

    awful:
        Simple but inefficient, slow or otherwise awful algorithms for
        generating primes and testing for primality. This module is
        provided only for pedagogical purposes (mostly as a lesson in
        what *not* to do).

    factors:
        Factorise small numbers into the product of primes.

    probabilistic:
        Generate and test for primes using probabilistic methods.

    sieves:
        Generate prime numbers using sieving algorithms.

    strategic:
        Various prime generating and testing functions implemented
        with the Strategy design pattern.

    utilities:
        Assorted utility functions.

plus the following private sub-modules:

    compat23:
        Internal compatibility layer, to support multiple versions of
        Python.

    tests:
        Unit and regression tests for the package.

The contents of the private sub-modules are subject to change or removal
without notice.

"""

from __future__ import division

import itertools
import warnings

import pyprimes.strategic
from pyprimes.compat23 import next
from pyprimes.utilities import filter_between


# Module metadata.
__version__ = "0.2.1a"
__date__ = "2015-01-07"
__author__ = "Steven D'Aprano"
__author_email__ = "steve+python@pearwood.info"


__all__ = ['is_prime', 'MaybeComposite', 'next_prime', 'nprimes',
           'nth_prime', 'prev_prime', 'prime_count', 'prime_partial_sums',
           'prime_sum', 'primes', 'trial_division',
          ]


class MaybeComposite(RuntimeWarning):
    """Warning raised when a primality test is probabilistic."""
    pass


# === Generate prime numbers ===

def primes(start=None, end=None):
    """Yield primes, optionally between ``start`` and ``end``.

    If ``start`` or ``end`` arguments are given, they must be integers.
    Only primes between ``start`` and ``end`` will be yielded:

    >>> list(primes(start=115, end=155))
    [127, 131, 137, 139, 149, 151]

    ``start`` is inclusive, and ``end`` is exclusive:

    >>> list(primes(5, 31))
    [5, 7, 11, 13, 17, 19, 23, 29]
    >>> list(primes(5, 32))
    [5, 7, 11, 13, 17, 19, 23, 29, 31]

    If ``start`` is not given, or is None, there is no lower limit and the
    first prime yielded will be 2. If ``end`` is not given or is None,
    there is no upper limit.
    """
    from pyprimes.sieves import best_sieve
    return pyprimes.strategic.primes(best_sieve, start, end)


def next_prime(n):
    """Return the first prime number strictly greater than n.

    >>> next_prime(97)
    101

    For sufficiently large n, over approximately 341 trillion, the result
    may be only probably prime rather than certainly prime.
    """
    return pyprimes.strategic.next_prime(is_prime, n)


def prev_prime(n):
    """Return the first prime number strictly less than n.

    >>> prev_prime(100)
    97

    If there are no primes less than n, raises ValueError.

    For sufficiently large n, over approximately 341 trillion, the result
    may be only probably prime rather than certainly prime.
    """
    return pyprimes.strategic.prev_prime(is_prime, n)


# === Primality testing ===

def is_prime(n):
    """Return True if n is probably a prime number, and False if it is not.

    >>> is_prime(103)
    True
    >>> is_prime(105)
    False


    For sufficiently large numbers, ``is_prime`` may be probabilistic rather
    than deterministic. If that is the case, ``is_prime`` will raise a
    ``MaybeComposite`` warning if ``n`` is only probably prime rather than
    certainly prime. The probability of a randomly choosen value being
    mistakenly identified as prime when it is actually composite is less
    than 1e-18 (1 chance in a million million million).

    If ``is_prime`` returns False, the number is certainly composite.
    """
    from pyprimes.probabilistic import is_probable_prime
    flag = pyprimes.strategic.is_prime(is_probable_prime, n)
    assert flag in (0, 1, 2)
    if flag == 2:
        message = "%d is only only probably prime" % n
        import warnings
        warnings.warn(message, MaybeComposite)
    return bool(flag)


def trial_division(n):
    """trial_division(n) -> True|False

    An exact but slow primality test using trial division by primes only.
    It returns True if the argument is a prime number, otherwise False.

    >>> trial_division(11)
    True
    >>> trial_division(12)
    False

    For large values of n, this may be slow or run out of memory.
    """
    return pyprimes.strategic.trial_division(primes, n)


# === Number theory convenience functions ===

def nprimes(n):
    """Convenience function that yields the first n primes only.

    >>> list(nprimes(10))
    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]

    """
    return itertools.islice(primes(), n)


def nth_prime(n):
    """nth_prime(n) -> int

    Return the nth prime number, starting counting from 1. Equivalent to
    p[n] (p subscript n) in standard maths notation.

    >>> nth_prime(1)  # First prime is 2.
    2
    >>> nth_prime(5)
    11
    >>> nth_prime(50)
    229

    """
    # http://oeis.org/A000040
    if n < 1:
        raise ValueError('argument must be a positive integer')
    return next(itertools.islice(primes(), n-1, n))


def prime_count(n):
    """prime_count(n) -> int

    Returns the number of prime numbers less than or equal to n.
    It is also known as the Prime Counting Function, π(x) or pi(x).
    (Not to be confused with the constant pi π = 3.1415....)

    >>> prime_count(20)
    8
    >>> prime_count(10780)
    1312
    >>> prime_count(30075)
    3251

    The number of primes less than n is approximately n/(ln n - 1).
    """
    # Values for pi(x) taken from here: http://primes.utm.edu/nthprime/
    # See also:  http://primes.utm.edu/howmany.shtml
    # http://mathworld.wolfram.com/PrimeCountingFunction.html
    # http://oeis.org/A000720
    return sum(1 for p in primes(end=n+1))


def prime_sum(n):
    """prime_sum(n) -> int

    prime_sum(n) returns the sum of the first n primes.

    >>> prime_sum(9)
    100
    >>> prime_sum(49)
    4888

    The sum of the first n primes is approximately n**2*(ln n)/2.
    """
    # See:  http://mathworld.wolfram.com/PrimeSums.html
    # http://oeis.org/A007504
    if n < 1:
        return 0
    return sum(nprimes(n))


def prime_partial_sums():
    """Yield the partial sums of the prime numbers.

    >>> p = prime_partial_sums()
    >>> [next(p) for _ in range(6)]  # primes 2, 3, 5, 7, 11, ...
    [0, 2, 5, 10, 17, 28]

    """
    # http://oeis.org/A007504
    n = 0
    for p in primes():
        yield n
        n += p

